### PR TITLE
Feature/commit and cleanup

### PR DESCRIPTION
**Pull request by alexeipab**
# Description

Adds flexible commit definitions that a separate from the push implementations. Enables multiple commit definitions. 
It also cleans up generics inheritance and error reporting of the data flow validation phases.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Unit tested.

### CODE REVIEW

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SimpleSparkDataFlow.scala [hunk 1]
```diff
@@ -5,43 +5,102 @@ import com.coxautodata.waimak.log.Logging
...
-                          , val tagState: DataFlowTagState = DataFlowTagState(Set.empty, Set.empty, Map.empty)) extends SparkDataFlow with Logging {
-
-  override protected def createInstance(in: DataFlowEntities, ac: Seq[DataFlowAction[SparkFlowContext]], tags: DataFlowTagState, schMeta: SchedulingMeta[SparkFlowContext]): DataFlow[SparkFlowContext] = {
-    // collect all labels that are inputs for SQL labels
-    val newSQLTables = sqlTables ++ ac.filter(_.getClass == classOf[SparkSimpleAction]).flatMap(a => a.asInstanceOf[SparkSimpleAction].sqlTables).toSet
-    new SimpleSparkDataFlow(spark, in, ac, newSQLTables, tempFolder, schMeta, commitLabels, tags)
-  }
+class SimpleSparkDataFlow(info: SimpleSparkDataFlowInfo) extends SparkDataFlow with Logging {
+
+//  override def addCommitLabel(label: String, definition: LabelCommitDefinition): SparkDataFlow = {
```
Comment (alexjbush) [inline waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SimpleSparkDataFlow.scala]: Remove commented code

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SparkInterceptors.scala [hunk 2]
```diff
@@ -30,7 +30,7 @@ object SparkInterceptors extends Logging {
   def addPostCacheAsParquet(sparkFlow: SparkDataFlow, outputLabel: String)
                            (dfFunc: Dataset[_] => Dataset[_])
                            (dfwFunc: DataFrameWriter[_] => DataFrameWriter[_]): SparkDataFlow = {
-    def post(data: Option[Any], sfc: SparkFlowContext): Option[Dataset[_]] = {
+    def post(data: Option[Any], sfc: FlowContext): Option[Dataset[_]] = {
```
Comment (alexjbush) [inline waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SparkInterceptors.scala]: You don't need sfc here?

You can use the flow context from the sparkflow object above then you don't need the cast below:

```scala
sfc.asInstanceOf[SparkFlowContext].spark.read.parquet(path)
```

becomes 
```scala
sparkFlow.flowContext.spark.read.parquet(path)
```

Comment (alexjbush) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SparkInterceptors.scala]: Then we could change the definition of:

```scala
sealed case class CachePostAction[T](run: (Option[T], FlowContext) => Option[T], override val labelToIntercept: String) extends PostAction[T](labelToIntercept)

sealed case class TransformPostAction[T](run: (Option[T], FlowContext) => Option[T], override val labelToIntercept: String) extends PostAction[T](labelToIntercept)
```

to
```scala
sealed case class CachePostAction[T](run: Option[T] => Option[T], override val labelToIntercept: String) extends PostAction[T](labelToIntercept)

sealed case class TransformPostAction[T](run: Option[T] => Option[T], override val labelToIntercept: String) extends PostAction[T](labelToIntercept)
```


This is a biggish change, but i don't see why it wouldn't work? Would tidy things up a little, however it wouldn't give users access to the spark context if they wanted to define custom interceptors anymore. What do you think?

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SimpleSparkDataFlow.scala (hunk 1 above)
Comment (alexeipab) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/SimpleSparkDataFlow.scala]: Fixed

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.coxautodata.waimak.dataflow
+
+package object spark {
+
+  type CleanUpStrategy[T] = (String, Seq[T]) => Seq[T]
```
[the remark this answers is not in the document]
Comment (alexeipab) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala]: Good idea, did changes, but thinking about it we could add type Label = String and use it everywhere.

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/DataFlow.scala [hunk 4]
```diff
@@ -463,10 +519,91 @@ case class SchedulingMeta[C](state: SchedulingMetaState, actionState: Map[String
...
+        if (committers.size != 1) Failure(new DataFlowException(s"Commit with name [${commit}] has ${committers.size} instead of 1"))
+        else loopTest(pushesToValidate.tail, committers.head.validate(dataFlow, commit, commits(commit)))
+      }
+    }
+
+    loopTest(pushes.keySet.intersect(commits.keySet), Success())
+  }
+
+  /**
+    * @param presentLabels
```
[the remark this answers is not in the document]
Comment (alexeipab) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/DataFlow.scala]: Fixed

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/DataFlow.scala [hunk 5]
```diff
@@ -16,11 +17,17 @@ import scala.util.{Failure, Success, Try}
   *
   * @tparam C the type of context which we pass to the actions
```
[the remark this answers is not in the document]
Comment (alexeipab) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/DataFlow.scala]: Fixed

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ParquetDataCommitter.scala [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+  /**
+    * Configures a default implementation of a cleanup strategy based on dates encoded into snapshot folder name.
+    *
+    * @param folderPrefix
+    * @param dateFormat
+    * @param numberOfFoldersToKeep
+    * @return
+    */
+  def dateBaseSnapshotCleanup(folderPrefix: String, dateFormat: String, numberOfFoldersToKeep: Int) =
```
Comment (alexjbush) [inline waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ParquetDataCommitter.scala]: Spelling of based

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexjbush) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala]: Big change, but that would be very nice

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ParquetDataCommitter.scala (hunk 6 above)
Comment (alexeipab) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/ParquetDataCommitter.scala]: Fixed

File: waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexjbush) [reply waimak-core/src/main/scala/com/coxautodata/waimak/dataflow/spark/package.scala]: Can do the label alias in another PR if you like

**Pull request merged**